### PR TITLE
Updating apiserver to return 202 when resource is being deleted asynchronously via cascading deletion

### PR DESCRIPTION
**Pull request by nikhiljindal**
As per https://github.com/kubernetes/kubernetes/issues/33196#issuecomment-278440622.

cc @kubernetes/sig-api-machinery-pr-reviews @smarterclayton @caesarxuchao @bgrant0607 @kubernetes/api-reviewers 

```release-note
Updating apiserver to return http status code 202 for a delete request when the resource is not immediately deleted because of user requesting cascading deletion using DeleteOptions.OrphanDependents=false.
```

### CODE REVIEW

**Comment by sttts** [comment]
Implementation looks fine. Any old clients in the wild which are broken by this? @smarterclayton

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 1]
```diff
@@ -867,12 +869,22 @@ func DeleteResource(r rest.GracefulDeleter, allowsOptions bool, scope RequestSco
...
+		status := http.StatusOK
+		// Return http.StatusAccepted if the resource was not deleted immediately and
+		// user requested cascading deletion by setting OrphanDependents=false.
+		// Note: We want to do this always if resource was not deleted immediately, but
+		// that will break existing clients.
+		// Other cases where resource is not instantly deleted are: namespace deletion
+		// and pod graceful deletion.
+		if !wasDeleted && options.OrphanDependents != nil && *options.OrphanDependents == false {
+			status = http.StatusAccepted
+		}
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Add a test case in apiserver for this.

**Comment by smarterclayton** [comment]
Yeah, this is more scoped to something new clients are asking for (client has to ask for orphan: false anyway) so it gets us in the right direction now.  We need to eventually return the estimate of time remaining and more data about the delete, but this is a minimal change to accomplish the goal.

**Comment by smarterclayton** [comment]
/approve

Add a test and this is LGTM

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 1 above)
Comment (nikhiljindal) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: @smarterclayton Any recommendation on what is the right file to put that test in?
I see `test/integration/auth/auth_test.go` being abused for status code testing. I can add a test there if there is no better place.

**Comment by caesarxuchao** [comment]
If orphanDependents is left nil, for any "controller" resources like RC/RS/Deployment, server will behave as if orphanDependents were set to "true" and will not immediately delete the resource. I understand the backward compatibility concern here, but it seems to be more confusing if the server sometimes returns 200 and sometimes 202 while the server behavior is the same.

**Comment by caesarxuchao** [comment]
Apart from the Orphan=nil example I gave, there are other cases where 202 *should* be returned but we cannot fix the apiserver in a backwards compatible way, e.g., if a user sets a finalizer for the object and then deletes the object. 
I think it will confuse users if apiserver sometimes, but not always, correctly choose between 200 and 202 in response to a deletion request. This concern is subjective so if @smarterclayton can approve this pull again I'll find peace with it.

**Comment by nikhiljindal** [comment]
I see this as if you use this new feature (by explicitly setting DeleteOptions.OrphanDependents) then apiserver returns detailed information (200 or 202), else it does not.
I agree that the best scenario would have been if we were able to always return 202 for asynchronous deletion irrespective of OrphanDependents but I also understand that we dont want to break existing clients.

I think we should fix this wherever we can instead of keeping it broken everywhere for consistency.

**Comment by caesarxuchao** [comment]
> I think we should fix this wherever we can instead of keeping it broken everywhere for consistency.

I think the opposite :) This is subjective. I'll be fine with it if @smarterclayton can approve the pull again.

**Comment by smarterclayton** [comment]
We have a few examples of this in other places, where we added new response types on new endpoints but preserved older behavior on older endpoints.  I think as we introduce new API groups and versions we'll want them to have the behavior added, and at some point we want clients to be able to opt in to "new" behavior on responses.  At a minimum, this incentivizes clients that have success iff 200 to deal with other values in the range, which reduces the potential for broken clients to propagate.

**Comment by smarterclayton** [comment]
The alternate representations of resources proposal contains way of allowing clients to signal they want "newer" API server behavior.

**Comment by nikhiljindal** [comment]
Pushed a new commit with an apiserver integration test.
@smarterclayton PTAL. Should be good to go now.

**Comment by nikhiljindal** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**